### ISSUE TITLE
Collaborate on parsers?

### ISSUE THREAD

**Issue by epage** [rustic-games/conventional-commit#1]
So there is the "parser" in clog (I say "parser" because it does some basic regexes to gather the subset of information clog needs) which is not reusable.  I went and made a reusable parser for [committed](https://github.com/crate-ci/committed) that I haven't yet broken out into a separate crate.  And there is now this one.

I figure it could be nice to see about working together on these efforts.

**Comment by JeanMertz** [comment]
Hey! Committed looks really nice. That seems like a handy tool to add to my workflow.

<sub>_(I tried using it locally, but unfortunately it didn't accept the `--style` flag for the CLI, I created a configuration file, but it complained about not capitalizing the first word of the subject, which isn't encouraged in conventional commit style, it is for any subsequent paragraphs though.)_</sub>

As for the parser, I'm all ears on collaborating on its implementation, moving to one parser and use that across all related tools. Whichever one you'd like to use, and wherever you'd like to host it (as long as it's on GitHub for ease of use, unfortunately I'm not a big fan of the GitLab UI), I'm good.

I pushed this up late last night to "ship something for the day", but I still have a host of issues to create to make progress on the parser, and also on [Jilu](https://github.com/rustic-games/jilu), the tool I wrote the parser for.

I'll continue to work through my list of things I'd like to get done, but feel free to ping me wherever you think it makes sense to work on the parser, or submit any PRs you think are relevant to the parser (I can give you the commit bit if you'd like to continue the work in this parser, just let me know).

The three biggest goals for me with this project are:

1. Correctness with the latest Conventional Commits specification (tested, perhaps some fuzzying, whatever gets the job done).
2. Easy to use API, conforming to the [Rust API Guidelines](https://rust-lang-nursery.github.io/api-guidelines/).
3. As few dependencies as possible, without sacrificing the first two points, even if it means a bit more boilerplate code.

**Comment by epage** [comment]
> (I tried using it locally, but unfortunately it didn't accept the --style flag for the CLI, I created a configuration file

Yes, I left `--style` off from [CLI args](https://github.com/crate-ci/committed/blob/master/docs/reference.md#configuration).  My thought was
- It'd be too complex to put everything in the CLI, forcing a config file
- The config file should support all things that do not change from run-to-run
- CLI args should contain things that do change from run to run

>  but it complained about not capitalizing the first word of the subject, which isn't encouraged in conventional commit style, it is for any subsequent paragraphs though.)

Keep in mind, `committed` is trying to be a general validator, even for people not using committed and it is generally recommended to capitalize.

As for committed, I'm not seeing anything to encourage or discourage it.  The examples are lower case but thats it.

> As for the parser, I'm all ears on collaborating on its implementation, moving to one parser and use that across all related tools. Whichever one you'd like to use, and wherever you'd like to host it (as long as it's on GitHub for ease of use, unfortunately I'm not a big fan of the GitLab UI), I'm good.

Well, let me look at my stuff and create whatever issues would be relevant and see if we align enough to merge efforts.  Personally, I'd prefer to welcome you over to the `crate-ci` org.  I have a couple hopes for that org
- Make processes and tools discoverable (as compared to the tons of conflicting blog posts when I started the CI for my rust projects)
- Increases the bus factor for projects by providing an area to host projects with people who have similar interests / skills (as compared to the generic bus factor projects)

>  I still have a host of issues to create to make progress on the parser, and also on Jilu, the tool I wrote the parser for.

So Jilu sounds like a competitor to clog-cli?  My automatic reaction is for us to also consider hosting that under `crate-ci` but I am also hesitant about "picking favorites" when multiple entries exist.  For azure pipelines stuff, we were able to get all the major players involved.  For this stuff, I haven't yet reached out to clog-cli's author.

> (I can give you the commit bit if you'd like to continue the work in this parser, just let me know).

As long as I'm not blocked on my stuff being reviewed and merged for too long, I'm fine either way.

> As few dependencies as possible, without sacrificing the first two points, even if it means a bit more boilerplate code.

This is my biggest area of concern because everyone has their own idea on where the line should be on this.  I had to walk away from a previous collaboration and do my own thing because the author was against deps in a confusing way (no `phf` and yet they accepted `clap` with its huge collection of indirect deps)

I'll give some examples of trade offs I've made
- I love `boolinator`; there have been times where using it would make the code clearer.  On the other hand, the dependency offers so little, I end up skipping over it.
- `derive_more` is something that I felt I could not justify using until recently, I saw that they added enough trait derivations that I'm now gladly pulling it into my projects (saving me a lot of `Display` and `From` boilerplate).  For example, see [report.rs](https://github.com/crate-ci/committed/blob/master/src/report.rs).  This will also be a big help when I move on from more prototype-y code and get rid of `failure::Error`.

**Comment by JeanMertz** [comment]
> Yes, I left `--style` off from [CLI args](https://github.com/crate-ci/committed/blob/master/docs/reference.md#configuration) [](/crate-ci/committed/blob/HEAD@{2019-08-10T14:41:10Z}/docs/reference.md#configuration)

Makes sense. I personally would've liked the flag, but I can see why you didn't add it.



> Keep in mind, `committed` is trying to be a general validator, even for people not using committed and it is generally recommended to capitalize.

True, but when a "style" is given, I would expect the tool to be configured for the "most common case" of that style. 

You are right that it's not explicitly defined in the spec (as are many other things, it's not that great of a spec yet, unfortunately), but the spec originates from the Angular project, which made this style of commit writing popular, [and they do define this](https://github.com/angular/angular/blob/22b96b96902e1a42ee8c5e807720424abad3082a/CONTRIBUTING.md#subject).

> So Jilu sounds like a competitor to clog-cli?

I wouldn't call it a _competitor_, more like a different take on the same subject, with some other philosophies guiding it.

> This is my biggest area of concern because everyone has their own idea on where the line should be on this.

Fair point. It _is_ a grey area, but it's also one actively being discussed in the Rust community. In general, I don't care if my binary tools have 100+ dependencies, I can build them once, and offer binaries for download. I do take care about which dependencies to pull in for library crates though, especially if they perform one simple task, that shouldn't force you to pull in 10+ other sub-dependencies. It's a balancing act, and I would never forgo using `serde` for example (although even there, there are some discussions in the community about it being fairly heavy, but that's mostly out of serde's control and simply a fact of the problem space it's solving).

But, to give you a more concrete example, the library currently has three dependencies:

```
$ cargo tree
conventional-commit v0.1.0
├── itertools v0.8.0
│   └── either v1.5.2
└── unicode-segmentation v1.3.0
```

We only use _one_ function provided by `itertools`, so it might be worth just copying that function over (it uses no unsafe, so I'm more comfortable in that case doing this), and reduce the dependency to only one crate (`unicode-segmentation`, which I would never think of not depending on, it's a vital part of providing correctness of the library's implementation).

_(this is leaving out the unfortunate effect dependencies have on local development speed, with the increase in compilation times for every new dependency added)_

In any case, I'm not stringent about it, but I have it on my radar when I develop in Rust, to try and find the right trade-offs.


>  My automatic reaction is for us to also consider hosting that under `crate-ci`

I'd say that would only be worth it if the crate got traction and turns out to be useful to others, and could use with a more shared collaboration. But, if there's any way I can help the `crate-ci` group advance their goals, I'm all ears, as I think it's an important (and undervalued) part of Rust's eventual success story.

**Comment by epage** [comment]
I've gone ahead and [asked about capitalization](https://github.com/conventional-commits/conventionalcommits.org/issues/176)

**Comment by JeanMertz** [comment]
Great!

By the way, I haven't pushed this yet (will do so tomorrow, I'm off to bed now), but I'm working on an update to this crate, in which I'm experimenting with the [Pest parser](https://pest.rs/).

I came up with the following grammar (so far):

```peg
commit  			= { SOI ~ message ~ EOI }
message 			= { header ~ body_block? ~ footer_block? }

header  			= { type_ ~ scope_block? ~ colon ~ space ~ description }
body_block 			= { blank_line ~ body }
footer_block		= { blank_line ~ footer }

body 				= { (!(footer_block | footer) ~ ANY)+ }
footer 				= { (trailer ~ NEWLINE?)+ }

type_       		= { ASCII_ALPHA+ }
scope_block 		= { open_paren ~ scope ~ close_paren }
scope 				= { ASCII_ALPHA+ }
description 		= { (!NEWLINE ~ ANY)+ }

trailer 			= { trailer_key ~ trailer_sep ~ space? ~ trailer_value }
trailer_key 		= { (!(trailer_sep | NEWLINE) ~ ANY)+ }
trailer_sep 		= { colon | pound }
trailer_value 		= { trailer_value_line ~ trailer_value_cont* }
trailer_value_line 	= { (!NEWLINE ~ ANY)+ }
trailer_value_cont 	= { NEWLINE ~ space+ ~ trailer_value_line }

open_paren 			= { "(" }
close_paren 		= { ")" }
colon 				= { ":" }
pound 				= { "#" }
space 				= { " " }
blank_line 			= { NEWLINE ~ NEWLINE }
```

It'll make it much easier to do correct parsing, and also support better error messages, such as:

```
 --> 1:6
  |
1 | feat(): my commit description
  |      ^---
  |
  = expected scope
```

Now, one thing that I did dug up during my research is that the `Co-Authored-By` lines in Git are actually called [_trailers_](https://git-scm.com/docs/git-interpret-trailers), and they already have a specification:

> trailers lines, that look similar to RFC 822 e-mail headers, at the end of the otherwise free-form part of a commit message.
> 
> ```
> token: value
> ```
> 
> This means that the trimmed <token> and <value> will be separated by ': ' (one colon followed by one space).

For a list of trailers used by git/linux related projects, [see the Git wiki "CommitMessageConventions" page](https://git.wiki.kernel.org/index.php/CommitMessageConventions).

In [RFC 822](https://tools.ietf.org/html/rfc822) there is also a section about [_folding long header fields_](https://tools.ietf.org/html/rfc822#section-3.4.8), which means the following are considered to be two trailers:

```
Trailer-1: hello world
Trailer-2: hello
 world
```

(having _at least_ one space in front of the next line makes it part of the previously defined trailer)

This is also part of the grammar I defined above.

So these examples are now parsed as expected:

```
feat: description
```

```
feat(scope): description
```

```
feat(scope): description

my body
```

```
feat(scope): description

my body

trailer 1: hello world
trailer 2: hello
 world
```

```
feat(scope): description

trailer 1: hello world
```

Anything else will produce a parsing error.

You can play with this grammar on the Pest [online editor](https://pest.rs/?bin=cwzb3#editor)

I suspect this could be of interest for the discussion in https://github.com/conventional-commits/conventionalcommits.org/issues/170. Specifically, I feel like swapping the current wording on `BREAKING CHANGE` to become inline with trailers (and so also requiring it to be in the footer, not anywhere else), would make the spec more clear and more in line with what Git already supports.

Feel free to add any remarks about this in the conventionalcommits.org issue, I won't be able to for at least another day or so.

I'll give you a ping when the new version is up.

**Comment by Mouvedia** [comment]
> `scope 				= { ASCII_ALPHA+ }`

@JeanMertz <del>I think you forgot to forbid whitespace in scopes.</del> It should be one word and permit `-` and `_`.

**Comment by JeanMertz** [comment]
Thanks for letting me know @Mouvedia. In fact, `ASCII_ALPHA` already only accepts `a..z` and `A..Z` characters, so spaces aren't allowed:

```
 --> 1:8
  |
1 | feat(my scope): my commit description
  |        ^---
  |
  = expected close_paren
```

About supporting `-` and `_`, can you point to where it mentions this syntax? All I can find is this:

> A scope MUST consist of a noun describing a section of the codebase surrounded by parenthesis

I guess if we were to accept [compound nouns](https://grammar.yourdictionary.com/parts-of-speech/nouns/compound-noun.html) we'd have to support both hyphens and spaces?

**Comment by Mouvedia** [comment]
If you don't accept `-` anyone using traincase will have to switch to camelcase. Just saying that it's probably too strict.

**Comment by JeanMertz** [comment]
> If you don't accept `-` anyone using traincase will have to switch to camelcase. Just saying that it's probably too strict.

Yeah, I agree with that sentiment, but the goal of a parser is to adhere to a spec, so we'd probably want the spec to be more clear on this.

Having said that, I'm going to change the parser to accept `-` and ` ` as valid characters for a scope, so that compound nouns are accepted. This can always be changed later if the spec becomes more clear on this. Thanks for the heads-up 👍.

**Comment by epage** [comment]
> By the way, I haven't pushed this yet (will do so tomorrow, I'm off to bed now), but I'm working on an update to this crate, in which I'm experimenting with the Pest parser.

Personally, I'm not much of a fan with pest after having used it.  It defers bugs to runtime because it gives you an enum of all of the rules when I tend to prefer APIs that catch bugs at runtime.

However, use what works for you.

> `header  			= { type_ ~ scope_block? ~ colon ~ space ~ description }`

Is this missing the use of `!` for tagging a breaking change within the subject?

> Now, one thing that I did dug up during my research is that the Co-Authored-By lines in Git are actually called trailers, and they already have a specification:

Yes, I wish the Conventional spec used the same terminology and syntax as existing git conventions.  Doesn't feel too conventional to me :/

> I suspect this could be of interest for the discussion in conventional-commits/conventionalcommits.org#170. Specifically, I feel like swapping the current wording on BREAKING CHANGE to become inline with trailers (and so also requiring it to be in the footer, not anywhere else), would make the spec more clear and more in line with what Git already supports.

I'm a little unclear on the concern with BREAKING CHANGE.  From my understanding, the footer is a less precisely defined trailer and either way you define it, `BREAKING CHANGE` in the spec is still compatible with it.

Also, not sure how a body starting with `BREAKING CHANGE` impacts things.

> Yeah, I agree with that sentiment, but the goal of a parser is to adhere to a spec, so we'd probably want the spec to be more clear on this.

Wish they'd provide a grammar to help resolve ambiguities.  For example, I was assuming `/` would be acceptable for hierarchical scopes like subsystems in the linux style of commit messages.

**Comment by JeanMertz** [comment]
I've since pushed several commits, such as 61e36288d88786a54af86f0048fb7ff5d4249c32, 0cae6b329ad1e5dffd7c57a853d6b9f686ca724b and some others to cover some of the issues raised here.

I'm closing this issue in favour of more specific issues already filed by @epage (thanks! ❤️) and discussions going on at the [conventionalcommits.org](https://github.com/conventional-commits/conventionalcommits.org) repository.

I welcome all contributions, and am still open to moving this repo some place else if that makes more sense, but for now I think the project is on a good trajectory to start off with.

Thanks for all the input, everyone, feel free to open new issues if there's anything still relevant not already covered in one of the recent commits, or described in an issue.
